### PR TITLE
Task #204040 learner distribution status for prerak view api

### PR DESCRIPTION
**Pull request by GitVivekHub**
# I have ensured that following `Pull Request Checklist` is taken care of before sending this PR

* [x] Code is formatted as per format decided
* [x] Updated acceptance criteria in tracker
* [x] Updated test cases in test-cases-tracker
* [x] Updated new API endpoint if any in common postman collection
* [x] Updated DB changes in db-tracker if any

### CODE REVIEW

File: src/src/facilitator/facilitator.service.ts [hunk 1]
```diff
@@ -1701,4 +1704,272 @@ export class FacilitatorService {
...
+					totalCount: count,
+					totalPages: totalPages,
+				},
+			});
+		} else {
+			return resp.status(200).json({
+				success: true,
+				message: 'Data found successfully!',
+				data: {
+					data: {},
```
Comment (parthgandhitekdi) [inline src/src/facilitator/facilitator.service.ts:1970]: data will be array.
Also send count and pages number.

**Review by parthgandhitekdi — CHANGES_REQUESTED** [review]

File: src/src/facilitator/facilitator.controller.ts [hunk 2]
```diff
@@ -119,5 +119,17 @@ export class FacilitatorController {
 		@Res() response: any,
 	) {
 		return this.facilitatorService.exportFileToCsv(request, body, response);
+
+
```
[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/facilitator/facilitator.controller.ts:123]: resolved

File: src/src/facilitator/facilitator.controller.ts [hunk 3]
```diff
@@ -119,5 +119,17 @@ export class FacilitatorController {
 		@Res() response: any,
 	) {
 		return this.facilitatorService.exportFileToCsv(request, body, response);
+
+		
+	}
+  
+	@Post('/admin/learner-status-distribution')
+	@UsePipes(ValidationPipe)
+	async getLearnerStatusDistribution(
```
[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/facilitator/facilitator.controller.ts:128]: resolved

File: src/src/facilitator/facilitator.service.ts [hunk 4]
```diff
@@ -1701,4 +1701,159 @@ export class FacilitatorService {
...
+
+		if (body.search && body.search !== '') {
+			let first_name = body.search.split(' ')[0];
+			let last_name = body.search.split(' ')[1] || '';
+
+			if (last_name?.length > 0) {
+				filterQueryArray.push(`{_or: [
+				{ user:{ first_name: { _ilike: "%${first_name}%" } } }
+				{user :{ last_name: { _ilike: "%${last_name}%" } } }
+				 ]} `);
```
[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/facilitator/facilitator.service.ts:1737]: resolved

File: src/src/facilitator/facilitator.service.ts [hunk 5]
```diff
@@ -1701,4 +1704,272 @@ export class FacilitatorService {
...
+		let filterQuery = '{ _and: [' + filterQueryArray.join(',') + '] }';
+
+		let variables = {
+			limit: limit,
+			offset: offset,
+		};
+
+		let qury = `query MyQuery($limit:Int, $offset:Int) {
+		program_faciltators(limit: $limit,
+			offset: $offset,where: ${filterQuery},order_by: ${sortType}) {
```
[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/facilitator/facilitator.service.ts:1778]: resolved

[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/facilitator/facilitator.service.ts:1778]: resolved

File: src/src/facilitator/facilitator.service.ts [hunk 6]
```diff
@@ -1701,4 +1704,272 @@ export class FacilitatorService {
...
+						status.filter((item) => item != 'identified'),
+					)}}},
+					{ status: { _is_null: true } }
+			 ]
+			}
+		)
+		{
+        aggregate {
+          count
+        }
```
[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/facilitator/facilitator.service.ts:1804]: resolved

File: src/src/facilitator/facilitator.service.ts [hunk 7]
```diff
@@ -1701,4 +1704,272 @@ export class FacilitatorService {
...
+					facilitator_first_name: facilitator.user.first_name,
+					facilitator_last_name: facilitator.user.last_name,
+					facilitator_id: facilitator.user.id,
+					facilitator_mobile_no: facilitator.user.mobile,
+					learnerData: learnerData,
+				};
+			});
+
+			let responseWithPagination = res.slice(offset, offset + limit);
+			const count = res[0].learnerData.length;
```
[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/facilitator/facilitator.service.ts:1953]: here i am taking count of total learners for the prerak, as i am already getting all the learners for the prerak through relation hence i am taking count of those learners present in the learnerData

File: src/src/facilitator/facilitator.service.ts [hunk 8]
```diff
@@ -1701,4 +1704,272 @@ export class FacilitatorService {
...
+				return {
+					facilitator_first_name: facilitator.user.first_name,
+					facilitator_last_name: facilitator.user.last_name,
+					facilitator_id: facilitator.user.id,
+					facilitator_mobile_no: facilitator.user.mobile,
+					learnerData: learnerData,
+				};
+			});
+
+			let responseWithPagination = res.slice(offset, offset + limit);
```
[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/facilitator/facilitator.service.ts]: resolved

File: src/src/facilitator/facilitator.service.ts [hunk 9]
```diff
@@ -1701,4 +1704,272 @@ export class FacilitatorService {
...
+				last_name: facilitator.user.last_name,
+				id: facilitator.user.id,
+				learner_total_count:
+					facilitator.learner_total_count.aggregate.count,
+				status_count: statusCount,
+			};
+		});
+
+		let responseWithPagination = res.slice(offset, offset + limit);
+		const count = res.length;
```
[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/facilitator/facilitator.service.ts]: resolved

File: src/src/facilitator/facilitator.service.ts [hunk 10]
```diff
@@ -1701,4 +1704,272 @@ export class FacilitatorService {
...
+			filterQueryArray.push(
+				`{user:{block:{_in: ${JSON.stringify(body?.block)}}}}`,
+			);
+		}
+
+		if (body.facilitator && body.facilitator.length > 0) {
+			filterQueryArray.push(
+				`{beneficiaries: {facilitator_id:{_in: ${JSON.stringify(
+					body.facilitator,
+				)}}}}`,
```
[the remark this answers is not in the document]
Comment (GitVivekHub) [reply src/src/facilitator/facilitator.service.ts]: resolved

**Review by parthgandhitekdi — APPROVED** [review]

**Review by parthgandhitekdi — CHANGES_REQUESTED** [review]
Need changes

**Review by parthgandhitekdi — CHANGES_REQUESTED** [review]

**Review by parthgandhitekdi — APPROVED** [review]

**Pull request merged**